lib/src/executable.dart

```diff
@@ -94,25 +104,142 @@ Future<void> handleFastFormat(List<String> args) async {
   }
 }
 
-void generateRunScript() {
-  if (shouldWriteRunScript) {
+void _deleteRunExecutableAndDigest() =>
+    [_paths.runExecutable, _paths.runExecutableDigest].forEach((p) {
+      final f = File(p);
+      if (f.existsSync()) f.deleteSync();
+    });
+
+/// Return true iff all the provided package names can be traced to 'hosted'
+/// entries in pubspec.lock.
+bool _allPackagesAreImportedImmutably(Iterable<String> packageNames) {
+  File pubspecLockFile = File('pubspec.lock');
+  if (!pubspecLockFile.existsSync()) return false;
+  final pubSpecLock = loadYamlDocument(pubspecLockFile.readAsStringSync());
+  return getDependencySources(pubSpecLock, packageNames)
+      .values
+      .every((source) => source == 'hosted' || source == 'git');
+}
+
+/// Return null iff it is not possible to account for all
+/// recompilation-necessitating factors in the digest.
+String? _computeRunScriptDigest() {
```

Comment:
It would have been nice to split this function up to call a config getter, configImports validator, and digest encoder function so it was easier to digest.